tests/main/snap-quota-cpu/task.yaml

```diff
@@ -0,0 +1,125 @@
+  # Bash cannot handle floating point numbers, so we multiple by 100 and use
+  # integer numbers instead
+  initialUsage=$(top -b -n 2 -d 0.2 | head | grep "stress" | awk '{print $9*100}')
+  if [ "$initialUsage" -lt 9000 ]; then
+    echo "Expected cpu usage to be atleast above 90%, got $initialUsage"
+    exit 1
+  fi
+
+  echo "Create a group with a snap in it"
+  snap set-quota group-one --cpu=2x25% test-snapd-stressd
+
+  echo "The systemd slice should be active now"
+  sliceName="snap.$(systemd-escape --path group-one).slice"
+  systemctl show --property=ActiveState "$sliceName" | MATCH "ActiveState=active"
+
+  echo "The service should also still be active"
+  snap services test-snapd-stressd.stress-sc | MATCH "test-snapd-stressd.stress-sc\s+disabled\s+active"
+
+  # systemd/kernel have three different locations for the cgroup pids depending
+  # on version
+  echo "The systemd slice should have two processes (all from stress snap) in it now"
+  cgroupsV1OldSystemdProcsFile="/sys/fs/cgroup/cpuacct/$sliceName/snap.test-snapd-stressd.stress-sc.service/cgroup.procs"
+  cgroupsV1ProcsFile="/sys/fs/cgroup/cpuacct/$sliceName/cgroup.procs"
+  cgroupsV2ProcsFile="/sys/fs/cgroup/$sliceName/snap.test-snapd-stressd.stress-sc.service/cgroup.procs"
+  if [ -e "$cgroupsV2ProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV2ProcsFile"
+  elif [ -e "$cgroupsV1OldSystemdProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV1OldSystemdProcsFile"
+  elif [ -e "$cgroupsV1ProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV1ProcsFile"
+  else
+      echo "cannot detect cgroup procs file"
+      exit 1
+  fi
+
+  #shellcheck disable=SC2016
+  retry --wait 1 -n 100 --env cgroupProcsFile="$cgroupProcsFile" sh -x -c 'test "$(wc -l < $cgroupProcsFile)" = 2'
+  SERVER_PID=$(cat "$cgroupProcsFile")
+
+  echo "And that process is the main PID for the stress daemon"
+  systemctl show --property=MainPID snap.test-snapd-stressd.stress-sc.service | MATCH "MainPID=$SERVER_PID"
+
+  echo "And the service is in the Control Group for the slice"
+  # using a regexp for the ControlGroup match here as on older systemd (16.04)
+  # the group is double escaped
+  systemctl show --property=ControlGroup snap.test-snapd-stressd.stress-sc.service | MATCH 'ControlGroup=/snap.group(.*)one.slice/snap.test-snapd-stressd.stress-sc.service'
+
+  # helper function that has retry logic to allow for cpu usage to stabilize
+  function ensure_cpu_usage_lower_than {
+    n=0
+    until [ "$n" -ge 5 ]
+    do
+      usage=$(top -b -n 2 -d 0.2 | head | grep "$1" | awk '{print $9*100}')
+      if [ "$usage" -lt "$2" ]; then
+        return 0
+      fi
+      n=$((n+1))
+      sleep 1
+    done
+
+    echo "Expected cpu usage to be maximum $2%, got $usage"
+    exit 1
+  }
+
+  echo "Verifying quota cpu usage, expecting around 50% usage"
+  # allow for up to 10% difference, this is necessary due to how we measure
+  ensure_cpu_usage_lower_than "stress" 6000
```

Comment:
I wonder how stable this will be. Perhaps we should just verify that the configuration has been set on the controller level and be done with it. The QA of the kernel mechanism can be done by the kernel folk.

Reply:
It usally settles at the exact percentage within 2-3 retries in ensure_cpu_usage_lower_than. The comment is actually a leftover before I wrote 'ensure_cpu_usage_lower_than', and it's perhaps not needed anymore to give it 10% difference. I was leaving it in to allow for swings if it should ever occur. I believe it's pretty stable. But I want to say that I'm definitely not against removing it if we don't actually care that the kernel quota works or not, it just makes things simpler for us

Reply:
It usally settles at the exact percentage within 2-3 retries in ensure_cpu_usage_lower_than. The comment is actually a leftover before I wrote 'ensure_cpu_usage_lower_than', and it's perhaps not needed anymore to give it 10% difference. I was leaving it in to allow for swings if it should ever occur. I can see that I might have to implement same retry for the rest of the measurements though. I'm definitely not against removing it if we don't actually care that the kernel quota works or not, it just makes things simpler for us.

Reply:
You know what, forget my comment, I can see even with my measurements I get see some instability over time. Maybe I should go ahead and remove them.